File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-8.js
class Solution {
    /**
     * @param {number[]} cost
     * @return {number}
     */
    minCostClimbingStairs(cost) {
        const memo = new Map();

        function dfs(i) {
            if (i >= cost.length) {
                return 0;
            };
            
            if (memo.has(i)) {
                return memo.get(i);
            };
            
            const result = cost[i] + Math.min(dfs(i + 1), dfs(i + 2));
            memo.set(i, result);
            return result;
        };

        return Math.min(dfs(0), dfs(1));
    }
}

File: Data Structures & Algorithms/word-break/submission-8.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        memo = {}

        def dfs(i, curr):
            if i in memo:
                return memo[i]
                
            if s[:len(curr)] != curr:
                return False
                
            if curr == s:
                return True

            if i >= len(s):
                return False

            for j in range(0, len(wordDict)):
                if dfs(i + len(wordDict[j]), curr + wordDict[j]):
                    memo[i] = True
                    return True
            
            memo[i] = False
            return False

        return dfs(0, "")